syntax = "proto3";

service ChatService {
    rpc CheckUserExists (UsernameRequest) returns (UserExistsResponse);
    rpc RegisterUser (RegisterRequest) returns (Response);
    rpc LoginUser (LoginRequest) returns (Response);
    rpc SendMessage (MessageRequest) returns (Response);
    rpc GetMessages (UsernameRequest) returns (MessagesResponse);
    rpc DeleteMessage (DeleteMessageRequest) returns (Response);
    rpc ListUsers (EmptyRequest) returns (UserListResponse);
    rpc SaveData (EmptyRequest) returns (Response);
}

message EmptyRequest {}

message UsernameRequest {
    string username = 1;
}

message UserExistsResponse {
    bool exists = 1;
}

message RegisterRequest {
    string username = 1;
    string password = 2;
}

message LoginRequest {
    string username = 1;
    string password = 2;
}

message MessageRequest {
    string sender = 1;
    string recipient = 2;
    string message = 3;
}

message MessagesResponse {
    repeated string messages = 1;
}

message DeleteMessageRequest {
    string username = 1;
    int32 message_id = 2;
}

message UserListResponse {
    repeated string users = 1;
}

message Response {
    string status = 1;
    string message = 2;
}